services/wallet/saved_addresses.go

```diff
@@ -15,41 +19,169 @@ type SavedAddress struct {
 	ChainID   uint64 `json:"chainId"`
 }
 
+type SavedAddressMeta struct {
+	Removed     bool
+	SyncClock   sql.NullInt64 // clock of the last sync
+	UpdateClock sql.NullInt64 // wall clock used to deconflict concurrent updates
+}
+
 type SavedAddressesManager struct {
 	db *sql.DB
 }
 
-func (sam *SavedAddressesManager) GetSavedAddresses(chainID uint64) ([]SavedAddress, error) {
-	rows, err := sam.db.Query("SELECT address, name, favourite, network_id FROM saved_addresses WHERE network_id = ?", chainID)
-	if err != nil {
-		return nil, err
-	}
-	defer rows.Close()
+func NewSavedAddressesManager(db *sql.DB) *SavedAddressesManager {
+	return &SavedAddressesManager{db: db}
+}
+
+const rawQueryColumnsOrder = "address, name, favourite, network_id, removed, sync_clock, update_clock"
 
-	var rst []SavedAddress
+// Retrieve raw data based on SELECT query using rawQueryColumnsOrder
+func getRawSavedAddressesFromDBRows(rows *sql.Rows) ([]SavedAddress, []SavedAddressMeta, error) {
+	var addresses []SavedAddress
+	var metas []SavedAddressMeta
 	for rows.Next() {
 		sa := SavedAddress{}
-		err := rows.Scan(&sa.Address, &sa.Name, &sa.Favourite, &sa.ChainID)
+		sam := SavedAddressMeta{}
```

Comment:
Any reason we are splitting the two in `SavedAddress` and `SavedAddressMeta`? 
it seems like it's adding complexity (we need to iterate over both and use it together), rather then having a single struct, and I'd have expected the data to be kept separate if in two different structs, so it could be counter-intuitive

Reply:
> Any reason we are splitting the two in SavedAddress and SavedAddressMeta?

I didn't want to pollute public APIs with internal metadata or even allow and have this separation of concerns. I also search for some kind of inheritance in go and try to use anonymous structures but didn't work. @cammellos how about having a higher struct that composes of both? Or do you prefer to have it simplified and expose to the API internal data?

> and I'd have expected the data to be kept separate if in two different structs, so it could be counter-intuitive

That was my first approach, to have a sync DB and build a more generalized sync framework and extend it to other structures, then realise that this will be a bigger change and premature given that I'm new to the code.

Reply:
> I didn't want to pollute public APIs with internal metadata, allow messing with sync/soft-delete functionality and have this separation of concerns.

It depends on what you mean, in general, you can use lower case fields, and those won't be accessible outside the namespace (you can see it as `private` fields essentially), they won't show in the JSON etc
That's probably common, @Samyoul is the in-house expert on what's the best way
I'd generally would not have 2 separate objects, but I'd be ok with either something like:
```
struct A {
 metadata metadatatype
 FieldA
 FieldB
}

struct metadatatype {
  metadataFileldA
  metadataFieldB
}
```
or even
```
struct A {

 FieldA
 FieldB
  metadataFileldA
  metadataFieldB
}

```

depending on usage and complexity of metadatatype

Reply:
You could use type embedding like @cammellos suggests, this is the closest thing to inheritance that go has.

```go
type metadatatype struct {
  metadataFileldA string
  metadataFieldB bool
}

type A struct {
 metadatatype
 FieldA int
 FieldB int
}
```

If you are concerned about polluting the API see the docs on this https://pkg.go.dev/encoding/json, unexported fields do not get marshalled into json. See example https://go.dev/play/p/NzXXXTwkRqb

If the fields need to be exported for whatever reason we can deal with that too:

```go
// Field is ignored by this package.
Field int `json:"-"
```

```go
type Metadatatype struct {
	MetadataFileldA string `json:"-"`
	MetadataFieldB  bool   `json:"-"`
}

type A struct {
	Metadatatype
	FieldA int
	FieldB int
}
```

See example here https://go.dev/play/p/564m4R-U5us , play around delete the json tags and see the result of marshalling.